api/download_ics.go

```diff
@@ -0,0 +1,104 @@
+package api
+
+import (
+	"errors"
+	"github.com/getsentry/sentry-go"
+	"github.com/gin-gonic/gin"
+	"github.com/joschahenningsen/TUM-Live/dao"
+	"github.com/joschahenningsen/TUM-Live/model"
+	"github.com/joschahenningsen/TUM-Live/tools"
+	"html/template"
+	"net/http"
+	"strconv"
+)
+
+func configGinDownloadICSRouter(router *gin.Engine) {
+	router.GET("/api/download_ics/:year/:term/:slug", downloadICS)
+}
+
+func downloadICS(c *gin.Context) {
+	templates, err := template.ParseFS(staticFS, "template/*.gotemplate")
+	if err != nil {
+		c.AbortWithStatus(http.StatusBadRequest)
+		return
+	}
+
+	foundContext, exists := c.Get("TUMLiveContext")
+	if !exists {
+		sentry.CaptureException(errors.New("context should exist but doesn't"))
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+	tumLiveContext := foundContext.(tools.TUMLiveContext)
+	if tumLiveContext.User == nil {
+		c.AbortWithStatus(http.StatusForbidden)
+		return
+	}
+
+	slug, term := c.Param("slug"), c.Param("term")
+	year, err := strconv.Atoi(c.Param("year"))
+	if err != nil {
+		c.AbortWithStatus(http.StatusBadRequest)
+		return
+	}
+
+	course, err := dao.GetCourseBySlugYearAndTerm(c, slug, term, year)
+	if err != nil {
+		c.AbortWithStatus(http.StatusBadRequest)
+		return
+	}
+
+	if len(course.Streams) == 0 {
+		_, err := c.Writer.WriteString("No lectures found!")
```

Comment:
Doesn't that break the ical? I suggest not showing the download button in this case.

Reply:
Makes sense, I am changing it.